File: Clients/JavaQGl/Q/src/com/qframework/core/LayoutAreaCards.java
/*
   Copyright 2012, Telum Slavonski Brod, Croatia.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
   
   This file is part of QFramework project, and can be used only as part of project.
   Should be used for peace, not war :)   
*/


package com.qframework.core;

class LayoutAreaCards extends LayoutArea {


    enum SubType {

        NONE,
        DECK,
        DECKBACK,
        HAND,
        INHAND,
        CARDS
    }
    SubType mSubType = SubType.NONE;
    float mModifier = 0;
    float mRandX = 0;
    float mRandY = 0;
    float mRandZ = 0;
    float mThickness = 0.0005f;
    
    public LayoutAreaCards(String subtype, GameonApp app) {
    	super(app);

        int val = subtype.charAt( subtype.length() -1 );
        if (val >= '0' && val <= '9')
        {
        	mModifier = (val-'0')*1.0f+1.0f;
        	mRandX = mModifier * 0.01f;
        	mRandY = mModifier * 0.01f;
        	mRandZ = mModifier * 0.01f;
        }
         
        if (subtype.startsWith("deckback")) {
            mSubType = SubType.DECKBACK;

        } else if (subtype.startsWith("deck")) {
            mSubType = SubType.DECK; 
        } else if (subtype.startsWith("hand")) {
            mSubType = SubType.HAND;
        } else if (subtype.startsWith("inhand")) {
            mSubType = SubType.INHAND;
        }else if (subtype.startsWith("cards")) {
            mSubType = SubType.CARDS;
        }


        mType = LayoutArea.Type.CARDS;
    }

    public void initLayout() {
        super.initLayout();

        if (mSubType == SubType.DECK) {
            initDeck(false);
        } else if (mSubType == SubType.DECKBACK) {
            initDeck(true);
        } else if (mSubType == SubType.HAND) {
            initHand();
        } else if (mSubType == SubType.INHAND) {
            initInHand();
        } else if (mSubType == SubType.CARDS) {
            initCards();
        }

    }

    void initDeck(boolean turnback)
    {
        int fieldc = 0;   	
        float width = mBounds[0];
        float height = mBounds[1];
        int fieldind = 0;        
        LayoutField field= null;
        for (int a=0; a< mSize ; a++)
        {
            setField(a);
            field = (LayoutField)mItemFields.elementAt(fieldind++);

            field.mX = 0;
            field.mY = 0;
            field.mW = 1;
            field.mH = 1;
            field.mZ = mThickness + a * mThickness;
            field.mRandX = mRandX;
            field.mRandY = mRandY;
            field.mRandRotZ = mRandZ * 360;
            if (turnback)
            {
                field.mItemRotY = 180;
            }
            fieldc++;
            
        }
        
        // adjust angle!
        if (this.mLayout == Layout.WEST_EAST)
        {
        	this.mRotation[2] = 90;	
        }else if (this.mLayout == Layout.EAST_WEST)
        {
        	this.mRotation[2] = 270;	
        }else if (this.mLayout == Layout.NORTH_SOUTH)
        {
        	this.mRotation[2] = 0;	
        }else if (this.mLayout == Layout.SOUTH_NORTH)
        {
        	this.mRotation[2] = 180;	
        }   
        
    }

    void initHand()
    {
        int fieldc = 0;
        // get field Width / height
        
        float width = mBounds[0];
        float divx = width - width / 2.2f;
        float height = mBounds[1];
        int fieldind = 0;
        float zadd = mThickness;
        for (int a=0; a< mSize ; a++)
        {
            setField(a);
            LayoutField field = mItemFields.get(fieldind++);
            field.mX = a * divx + divx/2 - width/2 + mRandX;
            field.mY = 0 + mRandY;
            field.mW = 1 - 1 / 2.2f;
            field.mH = 1;
            field.mZ = 0 + a * zadd;
            if (mRandZ > 0)
            {
            	field.mItemRotZ = (mRandZ-0.05f) * 360;
            }
            field.mRef.setPosition(field.mX,0,field.mZ);
            field.mRef.setScale(divx,1,1);
            fieldc++;
            
        }
    }

    void initInHand()
    {
    	int fieldc = 0;
	    float width = mBounds[0];
	    float divx = width / (float)mSize;
	    float divx2 = width / (float)mSizeW;
	    
	    float height = mBounds[1];
        int fieldind = 0;
        for (int a=0; a< mSize ; a++)
        {
            setField(a);
	        LayoutField field = mItemFields.get(fieldind++);        
            field.mX = a * divx + divx/2 - width/2;
            field.mW =  1 / (float)mSizeW;
            field.mH = 1;
            field.mZ = mThickness + a * mThickness;//cosy * 1/3;
            fieldc++;
            field.mRandRotZ = mRandZ * 360;
            field.mRef.setPosition(field.mX,0,field.mZ);
            field.mRef.setScale(divx2,1,1);

        }
        
        // adjust angle!
        if (this.mLayout == Layout.WEST_EAST)
        {
        	this.mRotation[0] = 45;	
        }else if (this.mLayout == Layout.EAST_WEST)
        {
        	this.mRotation[0] = 45;
        	this.mRotation[2] = 180;	
        }else if (this.mLayout == Layout.NORTH_SOUTH)
        {
        	this.mRotation[0] = 90;
        	this.mRotation[1] = 90;	
        }else if (this.mLayout == Layout.SOUTH_NORTH)
        {
        	this.mRotation[0] = 90;
        	this.mRotation[1] = 270;	
        }        	
        
    }


    void initCards()
    {
        int fieldc = 0;
        float width = mBounds[0];
        float divx = width / (float)mSize;
        float divx2 = 1 / (float)mSizeW;
        
        float height = mBounds[1];
        int fieldind = 0;
        float zadd = mThickness;

        for (int a=0; a< mSize ; a++)
        {
	        setField(a);
	        LayoutField field = mItemFields.get(fieldind++);        
            field.mX = a * divx + divx/2 - width/2;
            field.mW = 1 / (float)mSizeW;
            field.mH = 1;
            field.mZ = 0 + zadd + a * mThickness;
            field.mY = 0;           
            if (mRandZ > 0)
            {
            	field.mItemRotZ = (mRandZ-0.05f) * 360;
            }
            fieldc++;
            field.mRef.setPosition(field.mX,0,field.mZ);
            field.mRef.setScale(divx2,1,1);
        }
        
        // adjust angle!
        if (this.mLayout == Layout.WEST_EAST)
        {
        	this.mRotation[2] = 0;	
        }else if (this.mLayout == Layout.EAST_WEST)
        {
        	this.mRotation[2] = 180;	
        }else if (this.mLayout == Layout.NORTH_SOUTH)
        {
        	this.mRotation[2] = 90;	
        }else if (this.mLayout == Layout.SOUTH_NORTH)
        {
        	this.mRotation[2] = 270;	
        }           
    }
}

File: Clients/Html5Q/tutorials/qframework/LayoutAreaCards.js
/*
   Copyright 2012, Telum Slavonski Brod, Croatia.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
   
   This file is part of QFramework project, and can be used only as part of project.
   Should be used for peace, not war :)   
*/


function LayoutAreaCards(subtype , app)
{
	this.mApp = app;
    this.mID= undefined;
    this.mBounds = [1,1,1];
    this.mLocation = [0,0,0,0,0];
    this.mRotation = [0,0,0];
    this.mScale = [1,1,1];

    this.mType = LayoutArea_Type.NONE;
    this.mState = LayoutArea_State.VISIBLE;
    this.mInitState = LayoutArea_State.VISIBLE;
    this.mOwner = [];
    this.mParent =  undefined;
    this.mLayout = LayoutArea_Layout.NONE;
    this.mDisplay = GameonWorld_Display.WORLD;
    this.mItemFields = [];
    this.mText = undefined;
    this.mData = undefined;
    this.mSize = 0;
    this.mSizeW = 0;
    this.mSizeH = 0;
    this.mColorForeground = undefined;
    this.mColorForeground2 = undefined;
    this.mColorBackground = undefined;
    this.mColorBackground2 = undefined;
    this.mOnclick = undefined;
	this.mOnFocusLost = undefined;
	this.mOnFocusGain = undefined;
	this.mWorld = undefined;
	this.mPageId = undefined;
	
	this.mModel = undefined;// = new GameonModel();
    this.mModelBack = undefined;

	this.mSubType = LayoutAreaCards_SubType.NONE;
	
	this.mColors = new Array(4);
	this.mColors[0] = this.mApp.mColors.white;
	this.mColors[1] = this.mApp.mColors.white;
	this.mColors[2] = this.mApp.mColors.white;
	this.mColors[3] = this.mApp.mColors.white;
	
    this.mModifier = 0;
    this.mRandX = 0;
    this.mRandY = 0;
    this.mRandZ = 0;	

	var val = subtype.charAt( subtype.length -1 );
	if (val >= '0' && val <= '9')
	{
		this.mModifier = (val-'0')*1.0+1.0;
		this.mRandX = this.mModifier * 0.01;
		this.mRandY = this.mModifier * 0.01;
		this.mRandZ = this.mModifier * 0.01;
	}

	
	if (subtype.indexOf("deckback") == 0) {
		this.mSubType = LayoutAreaCards_SubType.DECKBACK;
	} else if (subtype.indexOf("deck") == 0) {
		this.mSubType = LayoutAreaCards_SubType.DECK;
	} else if (subtype.indexOf("hand") == 0) {
		this.mSubType = LayoutAreaCards_SubType.HAND;
	} else if (subtype.indexOf("inhand") == 0) {
		this.mSubType = LayoutAreaCards_SubType.INHAND;
	}else if (subtype.indexOf("cards") == 0) {
		this.mSubType = LayoutAreaCards_SubType.CARDS;
	}

		
	this.mType = LayoutArea_Type.CARDS;
	this.mPageVisible = false;
	this.mThickness = 0.0005;
	this.mActive = true;
	this.mSizeText = -1;
	this.mScrollers = [0,-0.5,0.5,0,0,0];
	this.mHasScrollH = false;
	this.mHasScrollV = false;
	this.mScollerAnim = undefined;
	this.mActiveItems = 1;
	
}

LayoutAreaCards.prototype = new LayoutArea();
LayoutAreaCards.prototype.constructor=LayoutAreaCards;


LayoutAreaCards_SubType = 
{
	NONE:0,
	DECK:1,
	DECKBACK:2,
	HAND:3,
	INHAND:4,
	CARDS:5		
}
	
// inheritance



LayoutAreaCards.prototype.initLayout = function() 
{
	if (this.mSubType == LayoutAreaCards_SubType.DECK) {
		this.initDeck(false);
	} else if (this.mSubType == LayoutAreaCards_SubType.DECKBACK) {
		this.initDeck(true);
	} else if (this.mSubType == LayoutAreaCards_SubType.HAND) {
		this.initHand();
	} else if (this.mSubType == LayoutAreaCards_SubType.INHAND) {
		this.initInHand();
	} else if (this.mSubType == LayoutAreaCards_SubType.CARDS) {
		this.initCards();
	}

}

LayoutAreaCards.prototype.initDeck = function(turnback)
{
	var fieldc = 0; 	
	var width = this.mBounds[0];
	var height = this.mBounds[1];
	var fieldind = 0;        
	var field= null;
	var rotzadd = 0;
	
	for (var a=0; a< this.mSize ; a++)
	{
		this.setField(a);
		field = this.mItemFields[fieldind++];

		x = field.mGridx;
		y = field.mGridy;
		field.mX = 0;
		field.mY = 0;
		field.mW = 1;
		field.mH = 1;
		field.mZ = this.mThickness + a * this.mThickness;
		field.mRandX = this.mRandX;
		field.mRandY = this.mRandY;
		field.mRandRotZ = this.mRandZ * 360;
		if (turnback)
		{
			field.mItemRotY = 180;
		}
		fieldc++;
	}
	
	if (this.mLayout == LayoutArea_Layout.WEST_EAST)
	{
		this.mRotation[2] = 90;	
	}else if (this.mLayout == LayoutArea_Layout.EAST_WEST)
	{
		this.mRotation[2] = 270;	
	}else if (this.mLayout == LayoutArea_Layout.NORTH_SOUTH)
	{
		this.mRotation[2] = 0;	
	}else if (this.mLayout == LayoutArea_Layout.SOUTH_NORTH)
	{
		this.mRotation[2] = 180;	
	}   
	
}

LayoutAreaCards.prototype.initHand = function()
{
	var fieldc = 0;
	// get field Width / height
	
	var width = this.mBounds[0];
	var divx = width - width / 2.2;
	var height = this.mBounds[1];
	var fieldind = 0;
	var zadd = this.mThickness;
	for (var a=0; a< this.mSize ; a++)
	{
		this.setField(a);
		var field = this.mItemFields[fieldind++];
		field.mX = a * divx + divx/2 - width/2 + this.mRandX;
		field.mY = 0 + this.mRandY;
		field.mW = 1 - 1 / 2.2;
		field.mH = 1;
		field.mZ = 0 + a * zadd;
		if (this.mRandZ > 0)
		{
			field.mItemRotZ = (this.mRandZ-0.05) * 360;
		}
		field.mRef.setPosition(field.mX,0,field.mZ);
		field.mRef.setScale(divx,1,1);		
		fieldc++;
		
	}
}

LayoutAreaCards.prototype.initInHand = function()
{
	var fieldc = 0;
	var width = this.mBounds[0];
	var divx = width / this.mSize;
	var divx2 = width / this.mSizeW;
	var height = this.mBounds[1];
	var fieldind = 0;
	for (var a=0; a< this.mSize ; a++)
	{
		this.setField(a);
		var field = this.mItemFields[fieldind++];        
		field.mX = a * divx + divx/2 - width/2;
		field.mW =  1 / this.mSizeW;
		field.mH = 1;
		field.mZ = this.mThickness  + a * this.mThickness;
		field.mY = 0;//this.mY  - a * 0.002;
		field.mRandRotZ = this.mRandZ * 360;
		field.mRef.setPosition(field.mX,0,field.mZ);
		field.mRef.setScale(divx2,1,1);
		fieldc++;
	}
	
	if (this.mLayout == LayoutArea_Layout.WEST_EAST)
	{
		this.mRotation[0] = 45;	
	}else if (this.mLayout == LayoutArea_Layout.EAST_WEST)
	{
		this.mRotation[0] = 45;
		this.mRotation[2] = 180;	
	}else if (this.mLayout == LayoutArea_Layout.NORTH_SOUTH)
	{
		this.mRotation[0] = 90;
		this.mRotation[1] = 90;	
	}else if (this.mLayout == LayoutArea_Layout.SOUTH_NORTH)
	{
		this.mRotation[0] = 90;
		this.mRotation[1] = 270;	
	} 	
}


LayoutAreaCards.prototype.initCards = function()
{
	var fieldc = 0;
	var width = this.mBounds[0];
	var divx = width / this.mSize;
	var divx2 = width / this.mSizeW;
	
	var height = this.mBounds[1];
	var fieldind = 0;
	var zadd = this.mThickness;
	
	for (var a=0; a< this.mSize ; a++)
	{
		this.setField(a);
		var field = this.mItemFields[fieldind++];        
		field.mX = a * divx + divx/2 - width/2;
		field.mW = 1 / this.mSizeW;
		field.mH = 1;
		field.mZ = 0 + zadd + a * this.mThickness;
		field.mY = 0;
		if (this.mRandZ > 0)
		{
			field.mItemRotZ = (this.mRandZ-0.05) * 360;
		}
		field.mRef.setPosition(field.mX,0,field.mZ);
		field.mRef.setScale(divx2,height,1);
		fieldc++;

	}
	
	if (this.mLayout == LayoutArea_Layout.WEST_EAST)
	{
		this.mRotation[2] = 0;	
	}else if (this.mLayout == LayoutArea_Layout.EAST_WEST)
	{
		this.mRotation[2] = 180;	
	}else if (this.mLayout == LayoutArea_Layout.NORTH_SOUTH)
	{
		this.mRotation[2] = 90;	
	}else if (this.mLayout == LayoutArea_Layout.SOUTH_NORTH)
	{
		this.mRotation[2] = 270;	
	}
	
}

